### PR TITLE
Sign all the Windows binaries we build.

### LINKED ISSUE #3146: Sign all the bundled commandline binaries that we compile ourselves
On Windows we are signing the packaged Electron app, but there are many bundled programs that aren't signed.

We obviously don't want to sign bits that we download from upstream releases, like `helm` or `kubectl`. But all the tools we build from source (e.g. `docker.exe`, `nerdctl-stub.exe`, ...) should be signed, to decrease the likelihood of malware scanners complaining about them.

### PR DESCRIPTION
**Pull request by ericpromislow**
Fixes #3146

This means using a whitelist for the four directories we care about:
* current directory (for `Rancher Desktop.exe`)
* resources/resources/win32  (`wsl-helper.exe`)
* resources/resources/win32/bin (the binaries in this directory that we build ourselves)
* resources/resources/win32/internal (all the binaries in this directory)

Signed-off-by: Eric Promislow <epromislow@suse.com>

### CODE REVIEW

File: scripts/lib/sign-win32.ts [hunk 1]
```diff
@@ -52,39 +52,47 @@ export async function sign(workDir: string) {
...
-
-  // make privileged-service.exe available to the instller during signing
+  const resourcesRootDir = 'resources/resources/win32';
+  const internalDir = path.join(resourcesRootDir, 'internal');
+  const binDir = path.join(resourcesRootDir, 'bin');
+  const whiteList: Record<string, Array<string>> = {
+    '.':                ['Rancher Desktop.exe'],
+    [resourcesRootDir]: ['wsl-helper.exe'],
+    [internalDir]:      ['host-resolver.exe', 'privileged-service.exe', 'steve.exe', 'vtunnel.exe'],
+    [binDir]:           ['docker.exe', 'docker-credential-none.exe', 'kuberlr.exe', 'nerdctl.exe', 'rdctl.exe'],
```
Comment (mook-as) [inline scripts/lib/sign-win32.ts:62]: We should add a comment that `docker.exe` is built by us, and therefore we need to sign it.

File: scripts/lib/sign-win32.ts [hunk 2]
```diff
@@ -52,39 +52,47 @@ export async function sign(workDir: string) {
...
+  };
+
+  // make privileged-service.exe available to the installer during signing
   const privilegedServiceFile = 'privileged-service.exe';
   const privilegedServiceFrom = path.join(unpackedDir, internalDir, privilegedServiceFile);
   const privilegedServiceTo = path.join(process.cwd(), 'resources/win32/internal', privilegedServiceFile);
 
   await fs.promises.copyFile(privilegedServiceFrom, privilegedServiceTo);
```
Comment (mook-as) [inline scripts/lib/sign-win32.ts:70]: We will need to sign this copy of `privileged-service.exe` too (it shows up on uninstall).

**Review by mook-as — CHANGES_REQUESTED** [review]
Unsigned files:

- `resources/win32/bin/docker-buildx.exe`
- `resources/win32/bin/docker-compose.exe`
- `resources/win32/bin/docker-credential-ecr-login.exe`
- `resources/win32/bin/helm.exe`
- `resources/win32/bin/kubectl.exe`

Of those, I think only `kubectl.exe` needs to be signed. (It's a copy of `kuberlr`.)

---
Note for myself in the future:
`ls -Recurse -Filter *.exe <thing> | Get-AuthenticodeSignature`

File: scripts/lib/sign-win32.ts (hunk 2 above)
Comment (jandubois) [reply scripts/lib/sign-win32.ts:70]: > kubectl.exe needs to be signed. (It's a copy of kuberlr.)

Does it? IIUC we are downloading the binary from Flavio's Github releases page, and don't build it with our CI. As a matter of policy we should only sign binaries where we have full control over the build and release process. Or am I missing something?

Why would we sign `kuberlr` but not `helm`?

Comment (mook-as) [reply scripts/lib/sign-win32.ts:70]: That's a fun question! I was assuming that "we" here was "SUSE", so it's still ours. If we decide that we _shouldn't_ sign it, then we shouldn't sign `kuberlr` either (and not just `kubectl`. Or we should build `kuberlr` ourselves.

As it is, though, it's at a minimum inconsistent.

Comment (jandubois) [reply scripts/lib/sign-win32.ts:70]: > we shouldn't sign kuberlr either (and not just kubectl. Or we should build kuberlr ourselves

Yes, I think we should sign neither of them (as one is just a copy of the other), and we should eventually build it ourselves.

>  I was assuming that "we" here was "SUSE"

I think it applies particularly to everything built by the processes controlled by the Rancher Desktop team.

Yes, Flavio works for SUSE, but it is his personal Github account that hosts it, so is this a release that should be signed by us? In that case you could argue that Matt also works for SUSE, so why not sign "his" helm releases as well?

It is not directly comparable, but I don't know how to draw a line clearly, which is why I would rather not sign something unless it is build directly by the project.

File: scripts/lib/sign-win32.ts (hunk 1 above)
Comment (ericpromislow) [reply scripts/lib/sign-win32.ts]: done

File: scripts/lib/sign-win32.ts (hunk 2 above)
Comment (ericpromislow) [reply scripts/lib/sign-win32.ts]: ensured the uninstaller version of `privileged-service.exe` is signed, pulled out `kuberlr.exe`, didn't add `kubectl.exe`

**Review by mook-as — APPROVED** [review]
LGTM; built installer and checked that all files are signed as expected.

**Pull request merged**